book/concepts.md

```diff
@@ -1,6 +1,131 @@
+>
+> Higher-level languages are being built on top of Hydroflow to generate 
+> distributed code in the form of multiple spinners. 
+> Meanwhile, you can use Hydroflow to write your own distributed code, by writing individual spinners that work together, 
+> and deploying them with a tool like [Hydroplane](https://github.com/hydro-project/hydroplane). See the [Hydro Ecosystem](./ecosystem.md) for more on this.
+
+### So how might a human write distributed systems with Hydroflow?
+As an illustration of how you can work at the Hydroflow layer, consider the 
+[Chat Server example](./example_8_chat_server.md). If you run that binary
+with the command-line argument `--role server` it will start a single spinner that is responsible for a chat server: receiving
+membership requests and messages from clients, and forwarding messages from individual clients to all other clients.
+If you run that binary with the argument `--role client` it will start a spinner that is responsible for a chat client, which 
+forwards chat messages from stdin to the server, and prints out messages sent by the server. As a distributed system, the chat 
+service would typically consist of many client spinners and a single server spinner.
+
+Note that this is an example of an extremely simple distributed system in a "star" or "hub-and spokes" topology: the multiple client spinners are completely independent of each other, and each talks only with the central server spinner. **A figure here would be nice**. If we wanted something more interesting, we could consider deploying a cluster of multiple server spinners, say for fault tolerance or geo-distribution. We'd need to change the code to take advantage of the many servers in a correct and reliable manner. But that's a topic for another section; for now, let's stay focused on the basics. 
+
+## The Life and Times of a Hydroflow Spinner
+Like most reactive services, we can envision a Hydroflow spinner running as an unbounded loop that is managed 
+by the runtime library. Each iteration of the spinner's loop is called a *tick*. Associated with the spinner is 
+a current *clock* value, which tells you how many ticks were executed by this spinner prior to the current tick. Each 
+spinner produces totally-ordered, sequentially increasing clock values, which you can think of as the "local logical time"
+at the spinner.
+
+The spinner's main loop works as follows:
+1. Ingest a batch of data from one or more inbound channels, and deliver them to the appropriate `source_xxx` operators in the Hydroflow spec.
+2. Execute the Hydroflow spec. If the spec has cycles, continue executing it until it reaches a "fixpoint" on the current batch; i.e. it no longer produces any new data anywhere in the flow. Along the way, any data that appears in an outbound channel is sent to the appropriate destination.
+3. Advance the local clock before starting the next tick.
+
+The spinner's main loop is shown in the following diagram:
+**Diagram goes here**
```

Comment:
Add "TODO" or "<!-- TODO -->" for ctrl+f

Reply:
Ok stopping here, just looking at the codegen part now

Reply:
thx